### PR TITLE
add preprocessing step

### PR DESCRIPTION
**Pull request by oberonia78**
This PR adds the preprocessing step into the DSWx-SAR workflow. 
The preprocessing includes
- reading ancillary data such as dem, hand, worldcover, and reference water and relocating them into the mosaic rtc grid. 
- applying filter to the rtc images.

### CODE REVIEW

File: src/dswx_sar/dummy_dswx_s1.py
Comment (seongsujeong) [inline src/dswx_sar/dummy_dswx_s1.py:1]: I think it would be helpful to add a docstring for this code, so the users can understand the difference between this code and `dswx_sar.py`.

File: src/dswx_sar/filter_SAR.py
Comment (seongsujeong) [inline src/dswx_sar/filter_SAR.py:28]:
```suggestion
    window = np.ones([winsize, winsize]) / (winsize * winsize)
```

File: src/dswx_sar/filter_SAR.py
Comment (seongsujeong) [inline src/dswx_sar/filter_SAR.py:42]:
```suggestion
def weightingarr(im, winsize, k=K_DEFAULT,
                 cu=CU_DEFAULT, cmax=CMAX_DEFAULT):
```

File: src/dswx_sar/filter_SAR.py
Comment (seongsujeong) [inline src/dswx_sar/filter_SAR.py:63]:
```suggestion
    window_mean, window_std = compute_window_mean_std(im, winsize=winsize) 
```

File: src/dswx_sar/filter_SAR.py
Comment (seongsujeong) [inline src/dswx_sar/filter_SAR.py:69]:
```suggestion
w_t_arr[ (ci > cu) & (ci < cmax) ] =\
        np.exp((-k * (ci[( ci > cu) & (ci < cmax)] - cu))
               / (cmax - ci[( ci > cu) & (ci < cmax)]))```

File: src/dswx_sar/filter_SAR.py
Comment (seongsujeong) [inline src/dswx_sar/filter_SAR.py:58]:
```suggestion
        mean array
```

File: src/dswx_sar/filter_SAR.py
Comment (seongsujeong) [inline src/dswx_sar/filter_SAR.py:52]:
```suggestion
        window size to compute the mean and std.
```

File: src/dswx_sar/filter_SAR.py
Comment (seongsujeong) [inline src/dswx_sar/filter_SAR.py:89]:
```suggestion
        filtered intensity image.
```

File: src/dswx_sar/filter_SAR.py
Comment (seongsujeong) [inline src/dswx_sar/filter_SAR.py:94]:
```suggestion

```

File: src/dswx_sar/filter_SAR.py
Comment (seongsujeong) [inline src/dswx_sar/filter_SAR.py:95]:
```suggestion
    w_t, mean, _ = weightingarr(img, win_size, k, cu, cmax)
```

File: src/dswx_sar/filter_SAR.py
Comment (seongsujeong) [inline src/dswx_sar/filter_SAR.py:100]:
```suggestion
```
Removing couple of whitespace here

File: src/dswx_sar/pre_processing.py
Comment (seongsujeong) [inline src/dswx_sar/pre_processing.py:21]: Can you check if the description in the docstring here makes sense? Thanks

File: src/dswx_sar/pre_processing.py
Comment (seongsujeong) [inline src/dswx_sar/pre_processing.py:38]: I think adding a docstring for the class `AncillaryRelocation` would be helpful to understand

File: src/dswx_sar/pre_processing.py
Comment (seongsujeong) [inline src/dswx_sar/pre_processing.py:5]:
```suggestion
import numpy as np
```

File: src/dswx_sar/pre_processing.py
Comment (seongsujeong) [inline src/dswx_sar/pre_processing.py:4]: Can you check if this module was used in this code? Thanks

File: src/dswx_sar/pre_processing.py
Comment (seongsujeong) [inline src/dswx_sar/pre_processing.py:72]: Description for the parameter `relocated_file_str` is missing here.

File: src/dswx_sar/pre_processing.py
Comment (seongsujeong) [inline src/dswx_sar/pre_processing.py:90]: Can you check if the function actually supposed to return something?

File: src/dswx_sar/dummy_dswx_s1.py
Comment (oberonia78) [reply src/dswx_sar/dummy_dswx_s1.py:1]: @seongsujeong `dummpy_dswx_s1.py` code will be removed in the future release. This code is made for interface delivery, but all lines implemented is simply using the fixed thresholds. Anyway, the new codes will be added. Please ignore the errors or missing docstring in `dummpy_dswx_s1.py`

**Comment by oberonia78** [comment]
@seongsujeong Thank you for your comments. It seems it requires more clearing up. I just finished fixing the code based on your comments. For your information, `dummpy_dswx_s1.py` will be replaced with the new files. So, I did not modify the code for this.

**Comment by seongsujeong** [comment]
@oberonia78 Thanks for the information. I'll disregard `dummpy_dswx_s1.py` when reviewing the PR.

File: src/dswx_sar/dswx_sar_util.py
Comment (seongsujeong) [inline src/dswx_sar/dswx_sar_util.py:64]:
```suggestion
        geotiff file path to read the band
    band_ind:
        Index of the band to read, starts from 0
```
Just a suggestion.

File: src/dswx_sar/mosaic_rtc_burst.py
Comment (seongsujeong) [inline src/dswx_sar/mosaic_rtc_burst.py:238]: Just a heads up, me and Gustavo are aware of some issue in `bursts_center` mode. You might need to update the related codes later on.

File: src/dswx_sar/pre_processing.py
Comment (seongsujeong) [inline src/dswx_sar/pre_processing.py:57]: AFAIK, `/science/LSAR/GCOV/grids/frequencyA/projection` seems to be from NISAR. At least I can tell this dataset does not exist in RTC-S1. Can you take a look?

File: src/dswx_sar/pre_processing.py
Comment (seongsujeong) [inline src/dswx_sar/pre_processing.py:101]: This function seems to to something complex. Please add a brief docstring to explain what it does.

File: src/dswx_sar/pre_processing.py
Comment (seongsujeong) [inline src/dswx_sar/pre_processing.py:114]:
```suggestion
            reftif = gdal.Open(ref_file, gdal.GA_ReadOnly)
```

File: src/dswx_sar/pre_processing.py
Comment (seongsujeong) [inline src/dswx_sar/pre_processing.py:127]: What is going to happen when `len(lat0) != ysize` is `True` but `len(lon0) != xsize` is `False`?

File: src/dswx_sar/pre_processing.py
Comment (seongsujeong) [inline src/dswx_sar/pre_processing.py:151]:
```suggestion
        gdal.Warp(output_tif_str, input_tif_str, options=opt)
```
In case you are not doing anything with `ds`

File: src/dswx_sar/pre_processing.py
Comment (seongsujeong) [inline src/dswx_sar/pre_processing.py:163]: I think I'm not following why `width*gt[4]` is added to `miny`, and `height*gt[2]` to `maxx`.
gt[4] and gt[2] would be zero as long as `gt` does not have rotation and shearing transformations.

File: src/dswx_sar/pre_processing.py
Comment (seongsujeong) [inline src/dswx_sar/pre_processing.py:155]:
```suggestion
        ds = gdal.Open(intput_tif_str, gdal.GA_ReadOnly)
```

File: src/dswx_sar/pre_processing.py
Comment (seongsujeong) [inline src/dswx_sar/pre_processing.py:199]:
```suggestion
    mosaic_prefix = processing_cfg.mosaic.mosaic_prefix
```

File: src/dswx_sar/pre_processing.py
Comment (seongsujeong) [inline src/dswx_sar/pre_processing.py:186]: Please check if `vegetation_cfg` is used in this function.

File: src/dswx_sar/pre_processing.py
Comment (seongsujeong) [inline src/dswx_sar/pre_processing.py:37]: A question. So `Relocation` here means resampling the data to the same geogrid as input RTC?

**Review by seongsujeong — COMMENTED** [review]
Here are my comments and suggestions. I have not pointed all the punctuations in this PR (whitespaces, lines, etc.) Maybe it would be a good idea to submit a separate PR after we're done with merging all important PRs.

File: src/dswx_sar/pre_processing.py
Comment (oberonia78) [reply src/dswx_sar/pre_processing.py:37]: Yes. Your understanding is correct.

File: src/dswx_sar/pre_processing.py
Comment (oberonia78) [reply src/dswx_sar/pre_processing.py:101]: I added the docstring. Thank you

**Comment by oberonia78** [comment]
@seongsujeong I revised the PR based on the comments. Let me know if you have any comments. If not, let's merge it ASAP.

File: src/dswx_sar/pre_processing.py
Comment (seongsujeong) [inline src/dswx_sar/pre_processing.py]:
```suggestion
                        np.min(lon0)-xspacing/2,
                        np.max(lon0)+xspacing/2]
```

Please check if my suggestion makes sense. I'll leave PEP8 issue for now.

File: src/dswx_sar/filter_SAR.py
Comment (seongsujeong) [inline src/dswx_sar/filter_SAR.py:73]: I think it would be more helpful to provide the DOI or reference information to the enhanced Lee filter.

**Review by seongsujeong — COMMENTED** [review]
My additional comments as below. More to follow - Seongsu

File: src/dswx_sar/mosaic_rtc_burst.py
Comment (seongsujeong) [inline src/dswx_sar/mosaic_rtc_burst.py:22]: Is this function to find the most frequent element in `num_list`? That being case. how about using `Counter` class in `collections`? Below is an example code:

```python
from collections import Counter

counter = Counter(num_list)
most_common = counter.most_common()
most_freq_elem = most_common[0][0]
```

File: src/dswx_sar/mosaic_rtc_burst.py
Comment (seongsujeong) [inline src/dswx_sar/mosaic_rtc_burst.py:111]:
```suggestion
    geotransform = [xcoord[0] - xres / 2, float(xres), 0,
                    ycoord[0] - yres / 2, 0, float(yres)]
```

There will be a half pixel offset between the first values of [`xcoord`, `ycoord`] and the upper left corner in the RTC products geotransformation (in geotiff). I would suggest to either apply the half pixel offset (like suggestion above), or load those geotransformation information from RTC layer in geotiff.

File: src/dswx_sar/mosaic_rtc_burst.py
Comment (seongsujeong) [inline src/dswx_sar/mosaic_rtc_burst.py:141]:
```suggestion
                     yRes=abs(yres),
```

**Review by seongsujeong — COMMENTED** [review]
Few more comments for this PR. I think this PR is almost ready to be merged.

File: src/dswx_sar/mosaic_rtc_burst.py
Comment (oberonia78) [reply src/dswx_sar/mosaic_rtc_burst.py:141]: Good point. Thanks for finding out.

File: src/dswx_sar/filter_SAR.py
Comment (oberonia78) [reply src/dswx_sar/filter_SAR.py:73]: Yes. I added the DOI and reference.

File: src/dswx_sar/pre_processing.py
Comment (oberonia78) [reply src/dswx_sar/pre_processing.py]: This part was made when RTC was stored in h5 file. So this case does not exist now, but I'll leave it for the NISAR mission. Of course, I changed the lines as you corrected.

**Comment by oberonia78** [comment]
@seongsujeong All bugs were corrected and docstrings were added.

**Review by seongsujeong — APPROVED** [review]
LGTM. There is [one comment](https://github.com/opera-adt/DSWX-SAR/pull/12/files#r1263019892) that was not take care of, but that is not a major issue to hold this PR from getting merged. - Seongsu

File: src/dswx_sar/pre_processing.py
Comment (seongsujeong) [reply src/dswx_sar/pre_processing.py]: What I wanted to mention is that `yspacing` is added to np.max(lon0) (line 132), which I thought we should add `xspacing`. Also it felt like that we should subtract `xspacing` from min(lon0) to get `W` which I think it means `West`

**Comment by seongsujeong** [comment]
one more comment [here](https://github.com/opera-adt/DSWX-SAR/pull/12#discussion_r1299219655) Please make sure to check that out before merging this PR.

File: src/dswx_sar/pre_processing.py
Comment (oberonia78) [reply src/dswx_sar/pre_processing.py]: Yes. You are right. Thank you for the reminders.

File: src/dswx_sar/dswx_sar_util.py
Comment (oberonia78) [reply src/dswx_sar/dswx_sar_util.py:64]: Thank you:)

File: src/dswx_sar/pre_processing.py
Comment (oberonia78) [reply src/dswx_sar/pre_processing.py]: This was observed when the RTC was stored in HDF5 file. We may need to review this part when we are going to deal with the NISAR RTC products.

**Pull request merged**
